src/Amadevus.RecordGenerator.Generators/RecordGenerator.cs

```diff
@@ -4,46 +4,70 @@
+                                  BuilderPartialGenerator.Instance,
+                                  DeconstructPartialGenerator.Instance);
+
         public RecordGenerator(AttributeData attributeData)
         {
             this.attributeData = attributeData;
         }
 
+        static readonly Task<SyntaxList<MemberDeclarationSyntax>> EmptyResultTask =
+            Task.FromResult(List<MemberDeclarationSyntax>());
+
         public Task<SyntaxList<MemberDeclarationSyntax>> GenerateAsync(TransformationContext context, IProgress<Diagnostic> progress, CancellationToken cancellationToken)
         {
-            var generatedMembers = SyntaxFactory.List<MemberDeclarationSyntax>();
-            var features = GetFeatures();
-
-            if (context.ProcessingNode is ClassDeclarationSyntax classDeclaration)
-            {
-                var descriptor = classDeclaration.ToRecordDescriptor(features);
-                generatedMembers = generatedMembers
-                    .AddRange(
-                        GenerateRecordPartials(descriptor)
-                        .Where(x => x != null));
-            }
-            return Task.FromResult(generatedMembers);
+            return context.ProcessingNode is ClassDeclarationSyntax cds
+                 ? GenerateAsync(cds)
+                 : EmptyResultTask;
 
-            IEnumerable<MemberDeclarationSyntax> GenerateRecordPartials(RecordDescriptor descriptor)
+            Task<SyntaxList<MemberDeclarationSyntax>> GenerateAsync(ClassDeclarationSyntax classDeclaration)
             {
+                var descriptor = classDeclaration.ToRecordDescriptor();
                 if (descriptor.Entries.IsEmpty)
+                    return EmptyResultTask;
+
+                var generatedMembers = new List<MemberDeclarationSyntax>();
+                var features = GetFeatures();
+                var partialKeyword = Token(SyntaxKind.PartialKeyword);
+
+                var partials =
+                    from g in PartialGenerators
+                    select g.Generate(descriptor, features)
+                    into g
+                    where g != null
+                    select
+                        ClassDeclaration(classDeclaration.Identifier.WithoutTrivia())
+                            .WithTypeParameterList(classDeclaration.TypeParameterList?.WithoutTrivia())
+                            .WithBaseList(g.BaseTypes.IsEmpty ? null : BaseList(SeparatedList(g.BaseTypes)))
+                            .WithModifiers(TokenList(g.Modifiers.Except(new[] {partialKeyword})
+                                .Append(partialKeyword)))
+                            .WithMembers(List(g.Members))
+                            .AddGeneratedCodeAttributeOnMembers();
+
+                foreach (var partial in partials)
                 {
-                    yield break;
+                    cancellationToken.ThrowIfCancellationRequested();
+                    generatedMembers.Add(partial);
```

Comment:
We could do this as `yield return/IEnumerable<MemberDeclarationSyntax>`. Return type would be simpler, exiting early would be just via `yield break`, no need for the List variable. Creating the SyntaxList would need to be done on the call site, but I don't think this'd hurt readability. Thoughts?

Reply:
There are a few problems with that. `yield break` means you can't test cancellation if we ever want to do that. Next, you can cancel a task but it's logically incorrect to cancel a iterable (`IEnumerable<MemberDeclarationSyntax>`). You'd have to change the return type to be `IEnumerator<MemberDeclarationSyntax>` because it's okay to cancel an _iteration_. You could still use `yield` but then then creating the final syntax list wouldn't be simple and would allocate a list per append.

Reply:
That's an extremely great explanation. Thank you.